tofurkey.8
.TH tofurkey 8 "Mar 2024" "v1.3.4"
.SH NAME
tofurkey \- Distributed sync for Linux TCP Fastopen key rotations
.SH SYNOPSIS
.SY tofurkey
.RB [ \-vno ]
.RB [ \-i\~\c
.IR seconds ]
.RB [ \-a\~\c
.IR /run/tofurkey.autokey ]
.RB [ \-k\~\c
.IR /path/to/main/secret ]
.YS
.SH DESCRIPTION
\fBtofurkey\fR is a simple daemon which manages the timely,
synchronized, deterministic rotation of TCP Fastopen (TFO) keys across a
cluster of Linux machines.  Keys are set via the Linux 5.10+ procfs
interface for dual keys in \fI/proc/sys/net/ipv4/tcp_fastopen_key\fR .

In order to synchronize across a cluster, it requires as input a
long-term main secret key file which is shared across the cluster (by
you and/or your configuration/secret management system), set via the
\fB\-k\fR argument, which points at a secret file with at least 32 bytes
of high-entropy random data to use as a main key.

If this is not supplied, a random secret will be auto-generated and
persisted to the (reboot-volatile) rundir by default, which will allow
for local TFO key rotation without any synchronization or reboot
persistence.  A warning will be issued to stderr at startup in this
case, as it's not expected to be a common operating mode in production.
.SH OPTIONS
.TP
\fB\-k\fR
Path to long-term main secret file generated and distributed to a
cluster by the administrator. This file must exist and have at least 32
bytes of secret high-entropy binary data, and will be re-read every
time TFO keys are generated. Mutually exclusive with \-a. If this option
is not provided, then the \-a autokey mode is the default.
.TP
\fB\-a\fR
Custom pathname to persist an auto-generated key, defaults to
\fI/run/tofurkey.autokey\fR. This file will be created if it's missing
and persisted across runs, but perhaps not across reboots at the
default path, and obviously affords no possibility of distributed
sync across a cluster. Mutually exclusive with \-k.
.TP
\fB\-i\fR
Interval seconds for key rotation, default is 86400, allowed range is
10 \- 604800, must be even. Daemon wakes up to rotate keys at every
half-interval of unix time to manage validity overlaps.
Intervals \fBmust\fR match across a cluster to get the same keys!
.TP
\fB\-v\fR
Verbose output to stderr
.TP
\fB\-n\fR
Dry-run mode - Data is not actually written to procfs, but everything
else still happens
.TP
\fB\-o\fR
One-shot mode - it will calculate the current keys and set them once
and then exit. Normal mode is to remain running and rotate keys on
timer intervals forever.
.SH TESTING OPTIONS
These options are for software testing and shouldn't be necessary in
normal use!  They're not shown in the normal CLI usage output.
.SY tofurkey
.RB [ ... ]
.RB [ \-V ]
.RB [ \-T\~\c
.IR n ]
.RB [ \-P\~\c
.IR /proc/sys/net/ipv4/tcp_fastopen_key ]
.YS
.TP
\fB\-V\fR
Print TFO keys, implies \-v, leaks short term secrets to stderr!
.TP
\fB\-T\fR
Set a fake unix time value, implies \-o, range 1e6 - 1e13
.TP
\fB\-P\fR
Override default procfs output path for setting keys
.SH MORE INFO
https://github.com/blblack/tofurkey


NEWS
1.3.4 - 2026-03-20
    * Now defends against the (unlikely) future possibility of EINTR during our
      long clock_nanosleep() waits, just in case.
    * Made some minor QA script tweaks and added more "const" to some fn args
    * Simplified the source tree layout

1.3.3 - 2026-03-20
    * Drop the Zig port entirely.  I wasn't keeping it up to date, and it was
      mostly an early playground for me learning some things about Zig a couple
      years ago. I love Zig now and use it in other things, but this tiny port
      no longer serves any purpose!

1.3.2 - 2024-04-22
    * Zig port: Drop support for 0.11.0 (0.12.0 is current)

1.3.1 - 2024-03-03
    * Zig port: tiny bugfix to allow building on 0.11.0 again

1.3.0 - 2024-03-03
    * Coredumps are disabled at startup via setrlimit()
    * Process address space locked into memory at startup via
      mlockall(), if running as root.
    * Various keys are now stored in local stack vars, rather than
      sodium_malloc allocations (the above changes bring most of the same
      benefits with less complexity).
    * Zig port is now pure Zig (no longer links libc or libsodium), but
      there's still a few bits in lsys.zig and getopt.zig that should be
      upstreamed to std in some form.

1.2.0 - 2024-02-27
    * TL;DR - No user-visible functional changes, but some log outputs
      changed and a ton of internal refactoring/cleanup happened
    * CLI Usage output reformatted
    * Several error messages improved
    * Log outputs to stderr are all prefixed by one of the strings: "info:",
      "error:", or "FATAL:"
    * More constraints checking at compile time
    * More error checking for unlikely failure modes and silly inputs
    * Lots of other misc internal refactoring/improvements
    * An EXPERIMENTAL port to Zig was added alongside the normal C version
      here in the same source tree.  See bottom of README.md for details.

1.1.0 - 2024-02-13
    * libev dependency removed.  The timing loop is now done manually, and
      just as correctly AFAIK, via clock_gettime() and clock_nanosleep().
      This makes everything simpler all around, but I wasn't confident in
      this mechanism until I explored the option a bit more!
    * Default interval changed from 6h to 24h.  Deploying this version will
      cause a rift in your keys if you're using the default interval.
    * Some stderr log outputs cleaned up a bit.
    * The CLI options used for software testing (-[VPT]) are no longer
      documented in README.md or the CLI usage output, but they are
      present in the man page in a separate section.  Also, -T now
      always implies -o.
    * The options -k and -a are now explicitly mutually exclusive.
    * Various internal refactorings to improve code quality, etc.
    * Documentation updated for clarity.

1.0.2 - 2024-02-07
    * Made the error messages for certain kinds of read() and write() failures
      less-confusing and more-informative.
    * Autokey file perms on creation changed from 0400 to 0600, to make it
      possible for it to truncate files it created in previous runs, in certain
      error cases.
    * Fixed the potential leak of a single fd in a corner-case error condition
      on startup in autokey mode.

1.0.1 - 2024-01-30
    * Minor fixups to manpage formatting
    * Minor source code formatting (wrap lines >= 100)
    * Some verbose (-v) stderr outputs changed in order to be a
      bit less confusing and/or over-verbose
    * A few pedantic fixups to improve the security of secret key
      data while loaded in memory for use

1.0.0 - 2024-01-24
    * Minor build improvements (dependencies, rundir stuff)
    * Minor code/documentation refactors/improvements
    * Example systemd unit gained a Documentation line and Restart=on-failure
    * Switching to semver-style versioning from here forward and calling it
      "stable".

0.9 - 2024-01-23
    * Minor fixup to Makefile CPPFLAGS handling, so that critical defines are
      not lost when package build systems set their own CPPFLAGS.

0.8 - 2024-01-23
    * Made "-k" optional rather than required.  When -k is not specified, the
      daemon attempts to auto-generate a local random key and persist it to
      the run directory for future use.  This mode of operation is only useful
      to get smoothly rotated keys on a single host; it cannot help you with
      synchronization across a cluster.
    * New option "-a" allows overriding the default path the autokey is stored
      to on disk (no effect if -k is used).
    * Moved the default installation directory for the main binary to
      /usr/sbin instead of /usr/bin
    * Updated example systemd unit to use sbin and no arguments, which means
      default autokeying.  This allows it to be deployed from distro packaging

src/tofurkey.zig
// SPDX-License-Identifier: 0BSD
// SPDX-FileCopyrightText: 2024 Brandon L Black <blblack@gmail.com>

const std = @import("std");
const builtin = @import("builtin");
const config = @import("config");
const posix = std.posix;
const os = std.os;
const log = std.log;
const crypto = std.crypto;
const assert = std.debug.assert;
const Allocator = std.mem.Allocator;
const getopt = @import("getopt.zig"); // Our POSIXy getopt() native impl
const lsys = @import("lsys.zig"); // non-libc Linux syscall stuff

// This whole project, by its functional nature, only works on Linux
comptime {
    if (builtin.os.tag != .linux) unreachable;
}

pub const std_options = .{
    .log_level = switch (builtin.mode) {
        .Debug => .debug,
        else => .info,
    },
};

// Constants for blake2b_kdf
const KDF_KEYBYTES = 32;

// The data length of a single ephemeral TFO key in binary form
const TFO_KEY_LEN = 16;

// The allocation length of the ASCII procfs form for 2x keys:
// pppppppp-pppppppp-pppppppp-pppppppp,bbbbbbbb-bbbbbbbb-bbbbbbbb-bbbbbbbb
// 64 bytes of hex digits plus 6x dashes and one comma = 71
const TFO_ASCII_ALLOC = 71;

// These timer "fudge" values are used to add an extra ~2.02s to the wake time
// for some insurance against various inaccuracies. Waking up slightly-late is
// fine, but waking up slight-early is not!
const FUDGE_S = 2;
const FUDGE_NS = 20_000_000; // 20ms

// Min/def/max user-specified interval values. While the upper bound is fairly
// arbitrary, the lower bound really should not go any lower than ~10s!  Given
// that the interval is cut in half for timers and there's a fixed 2.02s
// offset, this means with the min of 10, we'll be firing every 5 seconds with
// only ~3 seconds of room (after the offset) before the next firing. The time
// math also asserts on this minimum for functional reasons!
const MIN_IVAL = 10;
const DEF_IVAL = 86400;
const MAX_IVAL = 604800;

// Min/max real wall clock time we'll accept as legitimate and sane from
// clock_gettime(CLOCK_REALTIME) and also from our -T fake testing time. It's
// important that it can't go negative, that the min is larger than MAX_IVAL
// above, and that our time values (after adding an interval and fudge factor)
// can't saturate an i64)
const MIN_REAL_TIME = 1_000_000; // ~ Jan 12, 1970
const MAX_REAL_TIME = 100_000_000_000; // ~ Nov 16, 5138

comptime {
    // These must hold for any future changes of the constants above, or else
    // logic/correctness needs fixing in the actual code.
    assert(MIN_IVAL >= 10);
    assert(DEF_IVAL >= MIN_IVAL);
    assert(MAX_IVAL >= DEF_IVAL);
    assert(MIN_REAL_TIME > MAX_IVAL);
    assert(MAX_REAL_TIME + MAX_IVAL + FUDGE_S < std.math.maxInt(i64));
    // Assert that time_t (timespec.sec) can hold the same positive range as
    // i64. This code is intentionally not compatible with 32-bit time_t!
    assert(std.math.maxInt(posix.time_t) >= std.math.maxInt(i64));
}

// Default pathnames:
const def_autokey_path = config.rundir ++ "/tofurkey.autokey";
const def_procfs_path = "/proc/sys/net/ipv4/tcp_fastopen_key";

// non-secret context for the KDF (like an app-specific fixed salt)
const kdf_ctx = [8]u8{ 't', 'o', 'f', 'u', 'r', 'k', 'e', 'y' };

// Helpers to block/restore signals around sensitive code paths
fn block_all_signals() posix.sigset_t {
    var sigmask_prev: posix.sigset_t = undefined;
    posix.sigprocmask(posix.SIG.SETMASK, &os.linux.filled_sigset, &sigmask_prev);
    return sigmask_prev;
}

fn restore_signals(prev: posix.sigset_t) void {
    posix.sigprocmask(posix.SIG.SETMASK, &prev, null);
}

// Notify systemd iff NOTIFY_SOCKET was set in the environment. This allows
// systemd-level dependencies to work: if a network daemon binds/wants
// tofurkey, it can be assured the keys have been initially set.
fn sysd_notify_ready() !void {
    const spath = posix.getenv("NOTIFY_SOCKET") orelse return;
    // Must be an abstract socket or absolute path
    if (spath.len < 2 or (spath[0] != '@' and spath[0] != '/') or spath[1] == 0) {
        log.err("Invalid NOTIFY_SOCKET path '{s}'", .{spath});
        return error.SystemdNotify;
    }
    var sun = std.net.Address.initUnix(spath) catch |err| {
        log.err("Cannot create sockaddr struct for NOTIFY_SOCKET with path '{s}'", .{spath});
        return err;
    };
    if (sun.un.path[0] == '@')
        sun.un.path[0] = 0;

    const fd = posix.socket(posix.AF.UNIX, posix.SOCK.DGRAM | posix.SOCK.CLOEXEC, 0) catch |err| {
        log.err("Cannot create unix dgram socket fd for NOTIFY_SOCKET", .{});
        return err;
    };
    defer posix.close(fd);
    const strv = posix.sendto(fd, "READY=1", posix.MSG.NOSIGNAL, &sun.any, sun.getOsSockLen()) catch |err| {
        log.err("Cannot send READY=1 to NOTIFY_SOCKET '{s}'", .{spath});
        return err;
    };
    if (strv != 7) {
        log.err("Cannot send READY=1 to NOTIFY_SOCKET '{s}' (sent {d}/7 bytes)", .{ spath, strv });
        return error.SystemdNotify;
    }
}

// Safely read the main key file.
inline fn safe_read_key(mainkey: *[KDF_KEYBYTES]u8, mainkey_path: []const u8) !void {
    const key_fd = try posix.open(mainkey_path, .{ .ACCMODE = .RDONLY, .CLOEXEC = true }, 0);
    defer posix.close(key_fd);
    const readrv = try posix.read(key_fd, mainkey);
    if (readrv != mainkey.len)
        return error.KeyShortRead;
}

// Safely write the keys to procfs.
inline fn safe_write_procfs(keys_ascii: *const [TFO_ASCII_ALLOC]u8, procfs_path: []const u8) !void {
    const procfs_fd = try posix.open(procfs_path, .{ .ACCMODE = .WRONLY, .CLOEXEC = true, .SYNC = true }, 0);
    defer posix.close(procfs_fd);
    const writerv = try posix.write(procfs_fd, keys_ascii);
    if (writerv != keys_ascii.len)
        return error.KeyShortWrite;
}

// Safely write an autogenerated main key.
inline fn safe_write_autokey(mainkey: *const [KDF_KEYBYTES]u8, autokey_path: []const u8) !void {
    const autokey_fd = try posix.open(autokey_path, .{ .ACCMODE = .WRONLY, .CREAT = true, .TRUNC = true, .CLOEXEC = true, .SYNC = true }, posix.S.IRUSR | posix.S.IWUSR);
    defer posix.close(autokey_fd);
    const writerv = try posix.write(autokey_fd, mainkey);
    if (writerv != mainkey.len)
        return error.KeyShortWrite;
}

fn autokey_setup(mainkey_path: []const u8) !void {
    log.info("No -k arg given.  Will use a host-local autokey at {s}. This does not allow for cross-host synch!", .{mainkey_path});

    // Block signals while dealing with secure memory so that we always
    // wipe before exiting on a clean terminating signal
    const oldmask = block_all_signals();
    defer restore_signals(oldmask);

    // Temporary main key storage
    var mainkey: [KDF_KEYBYTES]u8 = undefined;
    defer crypto.utils.secureZero(u8, &mainkey);

    // Do a trial read to determine if there's an existing autokey file which
    // is usable. If it's not usable, invent a random key and persist it.
    safe_read_key(&mainkey, mainkey_path) catch {
        crypto.utils.secureZero(u8, &mainkey);
        log.info("Could not read autokey file {s}, generating a new random one", .{mainkey_path});
        crypto.random.bytes(&mainkey);
        try safe_write_autokey(&mainkey, mainkey_path);
    };
}

// Convert a pair of 16 byte raw binary keys to the ASCII hexadecimal format
// preferred by the Linux procfs interface
fn convert_keys_ascii(keys_ascii: *[TFO_ASCII_ALLOC]u8, kp: *const [TFO_KEY_LEN]u8, kb: *const [TFO_KEY_LEN]u8) !void {
    const h8 = "{x:0>2}{x:0>2}{x:0>2}{x:0>2}";
    const hk = h8 ++ "-" ++ h8 ++ "-" ++ h8 ++ "-" ++ h8;
    const hks = hk ++ "," ++ hk;
    // zig fmt: off
    const out = try std.fmt.bufPrint(keys_ascii, hks, .{
        kp[0],  kp[1], kp[2],  kp[3],  kp[4],  kp[5],  kp[6],  kp[7],
        kp[8],  kp[9], kp[10], kp[11], kp[12], kp[13], kp[14], kp[15],
        kb[0],  kb[1], kb[2],  kb[3],  kb[4],  kb[5],  kb[6],  kb[7],
        kb[8],  kb[9], kb[10], kb[11], kb[12], kb[13], kb[14], kb[15],
    });
    // zig fmt: on
    if (out.len != TFO_ASCII_ALLOC)
        return error.KeyFormattedLen;
}

// Must mirror libsodium:crypto_kdf_blake2b_derive_from_key() for compat
fn blake2b_kdf(out: *[TFO_KEY_LEN]u8, subkey: u64, key: *const [KDF_KEYBYTES]u8) void {
    var ctxpad: [16]u8 = [_]u8{0} ** 16;
    @memcpy(ctxpad[0..8], &kdf_ctx);
    var salt: [16]u8 = [_]u8{0} ** 16;
    std.mem.writeInt(u64, salt[0..8], subkey, .little);
    crypto.hash.blake2.Blake2b128.hash("", out, .{ .key = key, .salt = salt, .context = ctxpad });
}

// CLOCK_REALTIME, ignoring nanoseconds, range-validated, converted to u64
fn realtime_u64() !u64 {
    var ts = posix.timespec{ .sec = 0, .nsec = 0 };
    try posix.clock_gettime(.REALTIME, &ts);
    if (ts.sec < MIN_REAL_TIME or ts.sec > MAX_REAL_TIME) {
        log.err("Realtime value {d} is insane", .{ts.sec});
        return error.TimeRange;
    }
    return @intCast(ts.sec);
}

// timecalc() is a pure numeric function that takes care of the critical
// time-based calculations that are performed to create time-relevant keys.
// All inputs and outputs are u64s.  It takes in "now" (current unix wall time
// in seconds) and the configured interval, and outputs the two counter values
// for primary and backup TFO key derivation as well as the wall time we should
// next wake after in order to set keys again.
const tc_out = struct { ctr_primary: u64, ctr_backup: u64, next_wake: u64 };
fn timecalc(now: u64, interval: u64) tc_out {
    // "ctr_primary" is a counter number for how many whole intervals have
    // passed since unix time zero, and relies on the sanity checks here:
    assert(interval >= MIN_IVAL); // enforced in parse_args
    assert(now > interval); // enforced indirectly (cf comptime time asserts near top)
    const ctr_primary = now / interval;
    assert(ctr_primary > 0); // implicit, but maybe not compiler-obvious

    // "now_rounded_down" is the unix time exactly at the start of the current interval
    const now_rounded_down = ctr_primary * interval; // incapable of overflow, see above
    assert(now >= now_rounded_down); // implicit, but maybe not compiler-obvious

    // Which half is determined by the time that has passed since now_rounded_down
    const leftover = now - now_rounded_down;
    assert(leftover < interval); // implicit, but maybe not compiler-obvious
    const half_interval = interval / 2;
    const second_half = (leftover >= half_interval);

    // If we're past the middle of the interval, define "backup" as the next
    // upcoming key else define "backup" as the previous key (primary is always current):
    const ctr_backup = if (second_half) ctr_primary + 1 else ctr_primary - 1;

    // next time we should wake up after, which is the next round half-interval
    const to_add = if (second_half) interval else half_interval;
    const next_wake = now_rounded_down + to_add;

    return tc_out{
        .ctr_primary = ctr_primary,
        .ctr_backup = ctr_backup,
        .next_wake = next_wake,
    };
}

test "timecalc math checks" {
    try std.testing.expectEqualDeep(tc_out{ // A realistic example at MIN_IVAL
        .ctr_primary = 170901095,
        .ctr_backup = 170901094,
        .next_wake = 1709010955,
    }, timecalc(1709010952, 10));
    try std.testing.expectEqualDeep(tc_out{ // Maxing out time and ival
        .ctr_primary = 165343,
        .ctr_backup = 165344,
        .next_wake = 100_000_051_200, // realtime_u64() will bail
    }, timecalc(100_000_000_000, 604800));
    try std.testing.expectEqualDeep(tc_out{ // Max time, min ival
        .ctr_primary = 10_000_000_000,
        .ctr_backup = 9_999_999_999,
        .next_wake = 100_000_000_005, // realtime_u64() will bail
    }, timecalc(100_000_000_000, 10));
}

// Structure carrying fixed configuration from CLI down to functional parts
const cfg_s = struct {
    mainkey_path: []u8,
    procfs_path: []u8,
    fake_time: u64,
    interval: u64,
    verbose: bool,
    verbose_leaky: bool,
    dry_run: bool,
    one_shot: bool,

    fn init(alloc: Allocator) !*cfg_s {
        var self = try alloc.create(cfg_s);
        self.mainkey_path = undefined;
        self.procfs_path = undefined;
        self.fake_time = 0;
        self.interval = DEF_IVAL;
        self.verbose = false;
        self.verbose_leaky = false;
        self.dry_run = false;
        self.one_shot = false;
        try self._parse_args(alloc);
        return self;
    }

    fn _usage(comptime format: []const u8, args: anytype) noreturn {
        log.err(format, args);
        const usage_fmt =
            \\
            \\Usage: tofurkey [-vno] [-i seconds] [-a {s}] [-k /path/to/main/secret]
            \\  -k Path to long-term main secret file generated and distributed to a
            \\     cluster by the administrator. This file must exist and have at least
            \\     32 bytes of secret high-entropy binary data, and will be re-read every
            \\     time TFO keys are generated. Mutually exclusive with -a. If this option
            \\     is not provided, then the -a autokey mode is the default.
            \\  -a Custom pathname to persist an auto-generated key, defaults to
            \\     '{s}'. This file will be created if it's missing
            \\     and persisted across runs, but perhaps not across reboots at the
            \\     default path, and obviously affords no possibility of distributed
            \\     sync across a cluster. Mutually exclusive with -k.
            \\  -i Interval seconds for key rotation, default is {d}, allowed range
            \\     is {d} - {d}, must be even. Daemon wakes up to rotate keys at
            \\     every half-interval of unix time to manage validity overlaps.
            \\     Intervals *must* match across a cluster to get the same keys!
            \\  -v Verbose output to stderr
            \\  -n Dry-run mode - Data is not actually written to procfs, but everything
            \\     else still happens
            \\  -o One-shot mode - it will calculate the current keys and set them once
            \\     and then exit. Normal mode is to remain running and rotate keys on
            \\     timer intervals forever.
            \\
            \\This is tofurkey v1.3.2 (EXPERIMENTAL Zig variant)
            \\tofurkey is a tool for distributed sync of Linux TCP Fastopen key rotations
            \\More info is available at https://github.com/blblack/tofurkey
            \\
        ;
        const usage_args = .{ def_autokey_path, def_autokey_path, DEF_IVAL, MIN_IVAL, MAX_IVAL };
        std.io.getStdErr().writer().print(usage_fmt, usage_args) catch {};
        posix.exit(2);
    }

    fn _parse_args(self: *cfg_s, alloc: Allocator) !void {
        var arg_autokey: ?[*:0]const u8 = null;
        var arg_mainkey: ?[*:0]const u8 = null;
        var arg_procfs: ?[*:0]const u8 = null;
        var goi = getopt.getopt(os.argv, ":k:i:P:T:a:vVno");
        while (goi.next()) |optchar| {
            switch (optchar) {
                'v' => self.verbose = true,
                'n' => self.dry_run = true,
                'o' => self.one_shot = true,
                'k' => arg_mainkey = goi.getOptArg().?,
                'a' => arg_autokey = goi.getOptArg().?,
                'i' => {
                    const arg = goi.getOptArg().?;
                    const i = std.fmt.parseInt(u64, std.mem.span(arg), 10) catch {
                        _usage("Cannot parse '{s}' as u64", .{arg});
                    };
                    if (i < MIN_IVAL or i > MAX_IVAL or ((i & 1) != 0))
                        _usage("Interval value {d} is out of range or odd", .{i});
                    self.interval = i;
                },
                // These three are just for testsuite/debugging:
                'P' => arg_procfs = goi.getOptArg().?,
                'V' => {
                    self.verbose = true;
                    self.verbose_leaky = true;
                },
                'T' => {
                    const arg = goi.getOptArg().?;
                    const t = std.fmt.parseInt(u64, std.mem.span(arg), 10) catch {
                        _usage("Cannot parse '{s}' as u64", .{arg});
                    };
                    if (t < MIN_REAL_TIME or t > MAX_REAL_TIME)
                        _usage("Faketime value {d} is out of range", .{t});
                    self.fake_time = t;
                    self.one_shot = true;
                },
                // Error cases
                '?' => _usage("Invalid Option '-{c}'", .{goi.getOptOpt()}),
                ':' => _usage("Missing argument for '-{c}'", .{goi.getOptOpt()}),
                else => unreachable,
            }
        }

        if (goi.getOptInd() != os.argv.len)
            _usage("Excess unknown CLI arguments after options", .{});

        // Handle path string defaulting and autokey logic, etc
        if (arg_procfs) |arg| {
            self.procfs_path = try alloc.dupe(u8, std.mem.span(arg));
        } else {
            self.procfs_path = try alloc.dupe(u8, def_procfs_path);
        }

        if (arg_mainkey != null and arg_autokey != null)
            _usage("Cannot set both -k and -a", .{});

        var mainkey_path: [:0]const u8 = undefined;
        if (arg_mainkey != null) {
            mainkey_path = std.mem.span(arg_mainkey.?);
        } else if (arg_autokey != null) {
            mainkey_path = std.mem.span(arg_autokey.?);
        } else {
            mainkey_path = def_autokey_path;
        }

        if (mainkey_path.len == 0)
            _usage("Key path arguments cannot be empty strings", .{});
        if (mainkey_path[mainkey_path.len - 1] == '/')
            _usage("Key path arguments cannot have a trailing slash", .{});

        self.mainkey_path = try alloc.dupe(u8, mainkey_path);

        if (arg_mainkey == null)
            try autokey_setup(self.mainkey_path);
    }

    fn deinit(self: *const cfg_s, alloc: Allocator) void {
        alloc.free(self.mainkey_path);
        alloc.free(self.procfs_path);
        alloc.destroy(self);
    }
};

// The inner, security-sensitive part of set_keys()
fn set_keys_secure(cfg: *const cfg_s, now: u64, ctr_primary: u64, ctr_backup: u64) !void {
    // Block signals while dealing with secure memory so that we always wipe
    // before exiting on a clean terminating signal
    const oldmask = block_all_signals();
    defer restore_signals(oldmask);

    var key_ascii: [TFO_ASCII_ALLOC]u8 = undefined;
    defer crypto.utils.secureZero(u8, &key_ascii);
    {
        var key_primary: [TFO_KEY_LEN]u8 = undefined;
        var key_backup: [TFO_KEY_LEN]u8 = undefined;
        defer crypto.utils.secureZero(u8, &key_primary);
        defer crypto.utils.secureZero(u8, &key_backup);
        {
            var key_main: [KDF_KEYBYTES]u8 = undefined;
            defer crypto.utils.secureZero(u8, &key_main);
            try safe_read_key(&key_main, cfg.mainkey_path);
            blake2b_kdf(&key_primary, ctr_primary, &key_main);
            blake2b_kdf(&key_backup, ctr_backup, &key_main);
        }
        try convert_keys_ascii(&key_ascii, &key_primary, &key_backup);
    }

    if (cfg.verbose_leaky)
        log.info("Generated ASCII TFO keys for procfs write: [{d}] {s}", .{ now, &key_ascii });
    if (!cfg.dry_run)
        try safe_write_procfs(&key_ascii, cfg.procfs_path);
}

// Do the idempotent key generation + deployment based on current wall clock
// (even if it's not exactly when we would've woken up), then returns the next
// time we should wake up to rotate
fn set_keys(cfg: *const cfg_s, now: u64) !u64 {
    log.info("Setting keys for unix time {d}", .{now});
    const tc = timecalc(now, cfg.interval);
    try set_keys_secure(cfg, now, tc.ctr_primary, tc.ctr_backup);
    if (cfg.dry_run and cfg.verbose)
        log.info("Did not write to procfs because dry-run (-n) was specified", .{});
    return tc.next_wake;
}

pub fn main() !void {
    if (os.linux.geteuid() == 0)
        try lsys.mlockall(.{ .CURRENT = true, .FUTURE = true, .ONFAULT = true });
    const rlzero = posix.rlimit{ .cur = 0, .max = 0 };
    try posix.setrlimit(.CORE, rlzero);

    // plain GPA for config/path/etc
    var general_purpose_allocator = std.heap.GeneralPurposeAllocator(.{}){};
    const alloc = general_purpose_allocator.allocator();

    const cfg: *const cfg_s = try cfg_s.init(alloc);
    defer cfg.deinit(alloc);

    // Initially set keys to whatever the current wall clock dictates, and exit
    // immediately if one-shot mode
    const initial_now = if (cfg.fake_time != 0) cfg.fake_time else try realtime_u64();
    var next_wake = try set_keys(cfg, initial_now);
    if (cfg.one_shot) {
        log.info("Exiting due to one-shot mode (-o flag)", .{});
        return;
    }

    // For the long-running case, notify systemd of readiness after the initial
    // setting of keys above.
    try sysd_notify_ready();

    // We hang out in this time loop until something kills us
    if (cfg.verbose)
        log.info("Will set keys at each half-interval, when unix_time % {d} ~= 2", .{cfg.interval / 2});
    while (true) {
        const next_fudged = next_wake + FUDGE_S;
        if (cfg.verbose)
            log.info("Sleeping until next half-interval wakeup at {d}", .{next_fudged});
        try lsys.clock_nanosleep(.REALTIME, .ABSTIME, next_fudged, FUDGE_NS);
        next_wake = try set_keys(cfg, try realtime_u64());
    }
}

test {
    // Make sure we run unit tests in our other local import files
    std.testing.refAllDecls(getopt);
    std.testing.refAllDecls(lsys);
}
